K8sbykeshed/k8s-service-validator
# added tests for service sessionAffinity

Commits:
- added tests for service sessionAffinity
- refactored code structures for extendibility
- enabled validator to check if the connection reached target pod, instead of only check if connection to service is successful.

Goal:
- Verify session affinity functions, that same client send request to the service via same port, will reach to same destination pod
- Verify that same client send request to the service via **different** port, will reach to same destination pod

Executed test plan:
(setup)
1. add label `app=test-session-affinity` to pod-3 and pod-4
2. create service with same label, and sessionAffinity=ClientIP
(memorize affinity)
3. create from/to peer map to store from/to pods
4. let all four pods to reach session affinity service via port 80
5. save the destination pod for each into the from/to peer map
(verify via port 80)
6. create expected matrix based on from/to peer map
7. 4x4 matrix validation via clusterIP
(verify via port81)
8. do step 6 and 7
9. remove session affinity label from pod-3 and pod-4

result:
Session affinity service, from/to peers: map[pod-1:pod-3 pod-2:pod-4 pod-3:pod-3 pod-4:pod-3]
![image](https://user-images.githubusercontent.com/3346348/145538214-8d5109bc-32d6-49ca-8a12-19732107e715.png)
